plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -830,6 +840,108 @@ public void rollback()
+                .collect(toImmutableSet());
+
+        Map<ConnectorExpression, HiveApplyProjectionUtil.ProjectedColumnRepresentation> columnProjections = projectedExpressions.stream()
+                .collect(toImmutableMap(Function.identity(), HiveApplyProjectionUtil::createProjectedColumnRepresentation));
+
+        IcebergTableHandle icebergTableHandle = (IcebergTableHandle) handle;
+
+        // all references are simple variables
+        if (columnProjections.values().stream().allMatch(HiveApplyProjectionUtil.ProjectedColumnRepresentation::isVariable)) {
+            Set<ColumnHandle> projectedColumns = ImmutableSet.copyOf(assignments.values());
+            if (icebergTableHandle.getProjectedColumns().isPresent()
+                    && icebergTableHandle.getProjectedColumns().get().equals(projectedColumns)) {
+                return Optional.empty();
+            }
+            List<Assignment> assignmentsList = assignments.entrySet().stream()
+                    .map(assignment -> new Assignment(
+                            assignment.getKey(),
+                            assignment.getValue(),
+                            ((IcebergColumnHandle) assignment.getValue()).getType()))
+                    .collect(toImmutableList());
+            IcebergTableHandle newTableHandle = new IcebergTableHandle(
+                    icebergTableHandle.getSchemaName(),
+                    icebergTableHandle.getTableName(),
+                    icebergTableHandle.getTableType(),
+                    icebergTableHandle.getSnapshotId(),
+                    icebergTableHandle.getUnenforcedPredicate(),
+                    icebergTableHandle.getEnforcedPredicate(),
+                    Optional.of(projectedColumns));
+
+            return Optional.of(new ProjectionApplicationResult<>(
+                    newTableHandle,
+                    projections,
+                    assignmentsList,
+                    false));
+        }
+
+        ImmutableMap.Builder<String, Assignment> newAssignments = ImmutableMap.builder();
+        ImmutableMap.Builder<ConnectorExpression, Variable> newVariablesBuilder = ImmutableMap.builder();
+        ImmutableSet.Builder<ColumnHandle> projectedColumnsBuilder = ImmutableSet.builder();
+
+        for (Map.Entry<ConnectorExpression, HiveApplyProjectionUtil.ProjectedColumnRepresentation> entry : columnProjections.entrySet()) {
+            ConnectorExpression expression = entry.getKey();
+            HiveApplyProjectionUtil.ProjectedColumnRepresentation projectedColumn = entry.getValue();
+
+            // Create a new column handle
+            IcebergColumnHandle baseColumnHandle = (IcebergColumnHandle) assignments.get(projectedColumn.getVariable().getName());
+            IcebergColumnHandle projectedColumnHandle = createProjectedColumnHandle(baseColumnHandle, projectedColumn.getDereferenceIndices(), expression.getType());
+            String projectedColumnName = projectedColumnHandle.getQualifiedName();
```

Comment:
Is there a chance to get naming conflict here; get same `projectedColumnName` for different handles. It should not be possible if path elements and name cannot contain `dots`. It is a corner case, but still can you verify if that is possible?

Reply:
Looks like there's a validation for this in the Iceberg schema code:

```
trino:default> CREATE TABLE foo ("a.b" INT, a ROW (b INT));
Query 20210707_163341_00003_xkr2b failed: Invalid schema: multiple fields for name a.b: 1 and 3
org.apache.iceberg.exceptions.ValidationException: Invalid schema: multiple fields for name a.b: 1 and 3
```